tempesta_fw/tls.c

```diff
@@ -19,50 +19,250 @@
+		    return -EINVAL;
+	}	
+
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		const skb_frag_t *f = &skb_shinfo(skb)->frags[i];
+		if (tfw_tls_send_buf(tls, skb_frag_address(f), f->size))
+		    return -EINVAL;
+	}
+
+	kfree_skb(skb);
+
+	return 0;
+}
+
+/**
+ * Send @msg using TLS context @tls.
+ */
+int
+tfw_tls_send_msg(TfwTlsContext *tls, TfwMsg *msg)
+{
+	struct sk_buff *skb;
+
+	TFW_DBG("TLS:%p %s\n", tls, __func__);
+
+	while ((skb = ss_skb_dequeue(&msg->skb_list))) {
+		if (tfw_tls_send_skb(tls, skb)) {
+			kfree_skb(skb);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * Callback function which is called by TLS library.
+ */
+static int
+tfw_tls_send_cb(void *conn, const unsigned char *buf, size_t len)
+{
+	TfwConnection *c = conn;
+	TfwTlsContext *tls = c->tls;
+	struct sk_buff *skb = alloc_skb(MAX_TCP_HEADER + len, GFP_ATOMIC);
+
+	TFW_DBG("TLS:%p %s\n", tls, __func__);
+
+	if (unlikely(!skb))
+		return MBEDTLS_ERR_NET_SEND_FAILED;
+
+	skb_reserve(skb, MAX_TCP_HEADER);
+	skb_put(skb, len);
+
+	if (unlikely(skb_store_bits(skb, 0, buf, len)))
+		BUG();
+
+	ss_skb_queue_tail(&tls->tx_queue, skb);
+	ss_send(c->sk, &tls->tx_queue, SS_F_SYNC);
+
+	TFW_DBG2("TLS:%p %s -- ok (%lu bytes)\n", tls, __func__, len);
+
+	return len;
+}
+
+/**
+ * Callback function which is called by TLS library.
+ */
+static int
+tfw_tls_recv_cb(void *conn, unsigned char *buf, size_t len)
+{
+	TfwConnection *c = conn;
+	TfwTlsContext *tls = c->tls;
+	struct sk_buff *skb = ss_skb_peek_tail(&tls->rx_queue);
+
+	TFW_DBG("TLS:%p %s\n", tls, __func__);
+
+	if (unlikely(!skb))
+		return 0;
+
+	len = min_t(size_t, skb->len, len);
+	if (unlikely(skb_copy_bits(skb, 0, buf, len)))
+		BUG();
+
+	pskb_pull(skb, len);
+
+	if (unlikely(!skb->len)) {
+		ss_skb_unlink(&tls->rx_queue, skb);
+		kfree_skb(skb);
+	}
+
+	TFW_DBG2("TLS:%p %s -- ok (%lu bytes)\n", tls, __func__, len);
+
+	return len;
 }
 
-/* TODO Update @conn for newly established connection */
 static int
 tfw_tls_conn_init(TfwConnection *conn)
 {
+	int r;
+	TfwTlsContext *tls = conn->tls;
+
+	BUG_ON(!tls);
+
+	TFW_DBG("TLS:%p %s\n", tls, __func__);
+
+	mbedtls_ssl_init(&tls->ssl);
+
+	ss_skb_queue_head_init(&tls->rx_queue);
+	ss_skb_queue_head_init(&tls->tx_queue);
+
+	if (!tfw_tls.crt.version) {
+		TFW_ERR("TLS:%p no certificate specified\n", tls);
+		return -EINVAL;
+	}
+
+	r = mbedtls_ssl_setup(&tls->ssl, &tfw_tls.cfg);
+	if (r) {
+		TFW_ERR("TLS:%p setup failed (%x)\n", tls, -r);
+		return -EINVAL;
+	}
+
+	mbedtls_ssl_set_bio(&tls->ssl, conn,
+			    tfw_tls_send_cb, tfw_tls_recv_cb, NULL);
+
 	return 0;
 }
 
-/* TODO */
 static void
 tfw_tls_conn_drop(TfwConnection *conn)
 {
+	TfwTlsContext *tls = conn->tls;
+
+	TFW_DBG("TLS:%p %s\n", tls, __func__);
+
+	mbedtls_ssl_free(&tls->ssl);
+	tfw_http_conn_drop(conn);
+}
+
+static TfwMsg *
+tfw_tls_conn_msg_alloc(TfwConnection *conn)
+{
+	TFW_DBG("TLS:%p %s\n", conn->tls, __func__);
+
+	return tfw_http_conn_msg_alloc(conn);
```

Comment:
First of all, `conn_msg_alloc()` in TLS case is called when TLS message is arrived, that's not necessary encrypted HTTP message, but handshake message as well. In this particular case HTTP must not be bothered at all since we don't know which message we received. Rather, TSL FSM from ttls/ssl_tls.c must be activated here. Just see frang module for exmple.